[ADR] ADR-36: blockchain reactor refactor spec (#3506)
tendermint/tendermint: docs/architecture/adr-036-blockchain-reactor-refactor.md

responsibility. The heartbeat mechanism consists of periodically sending `StatusRequest` messages to peers and signalling to `Executor` in case a peer does not respond 
timely with the corresponding `StatusReport` message that a peer is not responsive so it should be removed from the peer list.    
This simpler architecture is easier to understand and makes writing of unit tests simpler tasks as most of the critical
logic is part of the `Controller` function. 

### Implementation changes

We assume the following system model for "fast sync" protocol: 

* a node is connected to a random subset of all nodes that represents its peer set. Some nodes are correct and some might be faulty. We don't make assumptions about ratio of  
  faulty nodes, i.e., it is possible that all nodes in some peer set are faulty. 
* we assume that communication between correct nodes is synchronous, i.e., if a correct node `p` sends a message `m` to a correct node `q` at time `t`, then `q` will receive
  message the latest at time `t+Delta` where `Delta` is a system parameter that is known. `Delta` is normally order of magnitude higher than the real communication delay 

>   (maximum) between correct nodes. Therefore if a correct node `p` sends a request message to a correct node `p` at time `t` and there is no the corresponding reply at time

Reviewer: `Therefore if a correct node p sends a request message to a correct node p` to `Therefore if a correct node p sends a request message to a correct node q`

Reviewer 2: ```suggestion
  (maximum) between correct nodes. Therefore if a correct node `p` sends a request message to a correct node `q` at time `t` and there is no the corresponding reply at time
```

>   (maximum) between correct nodes. Therefore if a correct node `p` sends a request message to a correct node `q` at time `t` and there is no the corresponding reply at time

Reviewer 3: there is no the corresponding reply at time
```suggestion
  (maximum) between correct nodes. Therefore if a correct node `p` sends a request message to a correct node `q` at time `t` and there is no corresponding reply at time
```

- `Correctness`: If a correct node `p` is connected to a correct node `q` for a long enough period of time, then `p` will eventually download all requested blocks from `q`.
- `Termination`: If a set of peers of a correct node `p` is stable (no new nodes are added to the peer set) for a long enough period of time, then protocol eventually terminates. 
- `Fairness`: A correct node `p` sends requests for blocks to all peers from its peer set.   

The `Controller` can be modelled as a function with clearly defined inputs:

* `State` - current state of the node. Contains data about connected peers and its behavior, pending requests, received blocks, etc.
* `Event` - significant events in the network.

producing clear outputs:

* `State` - updated state of the node,
* `Message` - signal what message to send,
* `TimeoutTrigger` - signal that timeout should be triggered.

```go
type Event int

const (
	EventUnknown Event = iota
	EventStatusReport
	EventBlockRequest
	EventNewBlock
	EventRemovePeer
	EventResponseTimeout
	EventTerminationTimeout
)

type BlockRequest struct {
	Height int64
	PeerID p2p.ID
}

type NewBlock struct {
	Height        int64
	Block         Block

> 	Commit        Commit

Reviewer 4: Is this last commit? Doesn't Block contain it already?

Author: To verify block at height h we need lastCommit from the block h+1.

> 	Commit        Commit

Reviewer 3: I think we need an overview of how all these data structures are used and when. Some control flow diagrams would be useful. 
Also, a part of this ADR should be the concurrency model/ design for scale/performance and testability. I believe this would be part of what you call the Executor task. I did some experiments with a version of the reactor poolRoutine that incorporates some of the Controller/ FSM changes. Sending one request per peer is slowing things down dramatically. Having hardcoded fixed limits may not be the best approach, maybe adaptive would work better. But in order to know, we need to have a good design (e.g. collect good stats, be able to abstract the things we sync, etc.) to understand what happens. Eyeing the logs is a nightmare. Not sure what we have today to collect operational data from nodes.

Author: I will try to add control diagram similar as we have for the current version. Maybe you can share source of the current diagram? Regarding performance concerns, extending this proposal to send multiple requests to a peer should be relatively straightforward. I will try to do it. I agree that number of pending requests per peer should be adaptive in ideal case but we need to estimate the complexity of such approach. We can maybe start with adding multiple pending requests (hardcoded value) to this proposal as this will let faster peer send multiple responses, and have them faster in position to serve new requests.

	PeerID        ID

> 	CurrentHeight int64

Reviewer 4: Rename to `CurrentPeerHeight`? Otherwise not clear what `CurrentHeight` refers to.

}

type StatusReport struct {
	PeerID ID
	Height int64
}

type RemovePeer struct {
	PeerID ID

> 	Height int64

Reviewer 3: Do we need Height here? In what conditions is this sent? Switch removes peer on error? Heartbeat loss?
Similar questions for the messages/ types below.

Author: We don't need Height, good point. It could be in case switch removes peer on error or in case StatusReport is not received on time.

}

type ResponseTimeout struct {
	PeerID ID
	Height int64
}

type TerminationTimeout struct {
	Height int64
}

type Message int

const (
	MessageUnknown Message = iota
	MessageBlockRequest
	MessageBlockResponse
	MessageStatusReport
)

type BlockRequestMessage struct {
	Height int64
	PeerID ID
}

> type BlockResponseMessage struct {

Reviewer 4: This seems to be the copy of `NewBlock` event? What's the reason for duplication?

Author: NewBlock is event that is created by the Executor upon receiving BlockResponseMessage.

	Height        int64
	Block         Block
	Commit        Commit
	PeerID        ID
	CurrentHeight int64
}

type StatusReportMessage struct {
	Height int64
}

type TimeoutTrigger int

const (
	TimeoutUnknown TimeoutTrigger = iota
	TimeoutResponseTrigger
	TimeoutTerminationTrigger
)
```

The Controller state machine can be in two modes (states): `FastSyncMode` when
it is trying to catch up with the network by downloading committed blocks,

> and `ConsensusMode` in which it executes Tendermint consensus protocol. We

Reviewer 4: Do we even need consensus mode in blockchain package?

Author: We should help other peers sync-up even during consensus mode. Idea here is to emphasise that fast-sync is a mode of operation and not one shot functionality (as it is in the current code).

assume that "fast sync" mode terminates once the Controller switch to
`ConsensusMode`.  

``` go
type State int

const (
	StateUnknown State = iota
	StateFastSyncMode
	StateConsensusMode
)

type ControllerState struct {
	Height             int64            // the first block that is not committed

> 	State              State            // mode of operation of the state machine

Reviewer 2: As this is part of `ControllerState` and the enum does have the suffix `Mode` already it might be beneficial streamline the naming to:

```go
type Mode int

const (
	ModeUnknown State = iota
	ModeFastSync
	ModeConsensus
)

type ControllerState struct {
	Height             int64            // the first block that is not committed
	Mode               Mode             // mode of operation of the state machine
	PeerMap            map[ID]PeerStats // map of peer IDs to peer statistics
	MaxRequestPending  int64            // maximum height of the pending requests
	FailedRequests     []int64          // list of failed block requests
	PendingRequestsNum int              // number of pending requests
	Store              []BlockInfo      // contains list of downloaded blocks
	Executor           BlockExecutor    // store, verify and executes blocks
}
```

	PeerMap            map[ID]PeerStats // map of peer IDs to peer statistics
	MaxRequestPending  int64            // maximum height of the pending requests

> 	FailedRequests     []int64          // list of failed block requests

Reviewer 4: why do we need this? for testing?

Author: If a request failed, then it is added to this list so it can be fetched from some other peer.

	PendingRequestsNum int              // number of pending requests
	Store              []BlockInfo      // contains list of downloaded blocks
	Executor           BlockExecutor    // store, verify and executes blocks
}

type PeerStats struct {
	Height         int64
	PendingRequest int64 // it can be list in case there are multiple outstanding requests per peer
}

type BlockInfo struct {
	Block  Block
	Commit Commit
	PeerID ID // a peer from which we downloading the corresponding Block and Commit
}
```

The `Controller` is initialized by providing initial height (`startHeight`) from which it will start downloading blocks from peers.

``` go

> func ControllerInit(state ControllerState, startHeight int64) ControllerState {

Reviewer 2: Why does init get an exisiting state? Is it not supposed to create a fresh state for us?

If so a more idiomatic name would be `NewControllerState`.

	state.Height = startHeight
	state.State = FastSyncMode
	state.MaxRequestPending = startHeight - 1
	state.PendingRequestsNum = 0
}

> func ControllerHandle(event Event, state ControllerState) (ControllerState, Message, TimeoutTrigger, Error) {

Reviewer 2: This could be more explicit:

``` go
handleEvent(state ControllerState, event Event) (ControllerState, Message, Timeout, Error)
```

Reviewer 3: As an alternative the FSM handle function could just return the new state and any error that occurred while processing the event from old state. Any actions (like sending a message or triggering a timeout) would be part of the handle implementation (action functions). Where needed they can make calls to interface functions where we can plug in tester or real reactors. 

I was also wondering if instead or along with the code we could have a diagram or description on what happens  on peer removal and update. There is also a case currently where we detect slow peers. We should also put in place the design to have N pending requests, I don't think 1 will be enough. This complicates things quite a bit unfortunately.

	msg = nil
	timeout = nil
	error = nil

	switch state.State {
	case StateConsensusMode:
		switch event := event.(type) {
		case EventBlockRequest:
			msg = createBlockResponseMessage(state, event)
			return state, msg, timeout, error
		default:

> 			error = "Only respond to BlockRequests while in ConsensusMode!"

Reviewer 3: Just a note that this is not the case with current implementation.

			return state, msg, timeout, error
		}

	case StateFastSyncMode:
		switch event := event.(type) {
		case EventBlockRequest:
			msg = createBlockResponseMessage(state, event)

> 			return state, msg, timeout, error

Reviewer 3: Here we are not in consensus mode and we reply to Block response messages. Also, this (the fast sync helper functionality) could be done outside the controller.

				}
			}
			state.PeerMap[event.PeerID] = peerStats
			return state, msg, timeout, error

		case EventRemovePeer:
			if _, ok := state.PeerMap[event.PeerID]; ok {
				pendingRequest = state.PeerMap[event.PeerID].PendingRequest
				if pendingRequest != -1 {
					add(state.FailedRequests, pendingRequest)
				}
				state.PendingRequestsNum--
				delete(state.PeerMap, event.PeerID)
				if state.PeerMap.isEmpty() || state.PendingRequestsNum == 0 {
					timeout = TerminationTimeoutTrigger{state.Height, TerminationTimeout}
				}
			} else {
				error = "Removing unknown peer!"
			}
			return state, msg, timeout, error

		case EventNewBlock:
			if state.PeerMap[event.PeerID] {
				peerStats = state.PeerMap[event.PeerID]
				if peerStats.PendingRequest == event.Height {
					peerStats.PendingRequest = -1
					state.PendingRequestsNum--
					if event.CurrentHeight > peerStats.Height {
						peerStats.Height = event.CurrentHeight
					}
					state.Store[event.Height] = BlockInfo{event.Block, event.Commit, event.PeerID}
					state = verifyBlocks(state) // it can lead to event.PeerID being removed from peer list

					if _, ok := state.PeerMap[event.PeerID]; ok {

> 						msg = createBlockRequestMessage(state, event.PeerID, peerStats.Height)

Reviewer: Previous blockchain reactor will send multi block requests in the same time, but new version only send new request until last block have received, will it slow down?

Author: With the new version, there is one pending request per peer. After a response (block) arrives, we will try to send a new request to the peer. This makes protocol more reactive towards responsive peers as the peers that send responses faster will be asked more. Having multiple pending requests (especially big number of requests as in the current version) is problematic as it opens the space for DDoS attacks. It is also not clear that having many pending requests is indeed more efficient. We will need to measure this. The main goal of this ADR is correctness and improved testability, and this should hopefully make performance improvements easier after.

						timeout = TimeoutTrigger{state.Height, TerminationTimeout}
					}
				}
			}

			return state, msg, timeout, error

		case EventTerminationTimeout:
			if state.Height == event.Height && state.PendingRequestsNum == 0 {
				state.State = ConsensusMode
			}
			return state, msg, timeout, error

		default:
			error = "Received unknown event type!"
			return state, msg, timeout, error
		}

	}
}

func createBlockResponseMessage(state ControllerState, event BlockRequest) BlockResponseMessage {
	msg = nil
	if _, ok := state.PeerMap[event.PeerID]; !ok {
		peerStats = PeerStats{-1, -1}
	}
	if state.Executor.ContainsBlockWithHeight(event.Height) && event.Height > peerStats.Height {
		peerStats = event.Height
		msg = BlockResponseMessage{
			Height:        event.Height,
			Block:         state.Executor.getBlock(eventHeight),
			Commit:        state.Executor.getCommit(eventHeight),
			PeerID:        event.PeerID,
			CurrentHeight: state.Height - 1,
		}
	}
	state.PeerMap[event.PeerID] = peerStats
	return msg
}

func createBlockRequestMessage(state ControllerState, peerID ID, peerHeight int64) BlockRequestMessage {
	msg = nil

> 	blockNumber = -1

Reviewer 4: rename to `BlockHeight`?

	// exist request r in state.FailedRequests such that r <= peerHeight

> 	if true {

Reviewer 4: ???

Author: Fixed in the latest version.

		blockNumber = r
		delete(state.FailedRequests, r)
	} else if state.MaxRequestPending < peerHeight {

> 		state.MaxRequestPending++

Reviewer 4: shouldn't we set it to `peerHeight`?

Reviewer 4: also, I can't find the place where we decrease it after the peer with the max height disconnects.

Author: There is no notion of max peer height in this proposal (therefore it is never reset). Termination mechanism is different (based on number of pending requests, size of the peer set and termination timeout). I have added more comments in the latest version. MaxRequestPending is the maximum height for which request is made to some peer.

		blockNumber = state.MaxRequestPending // increment state.MaxRequestPending and then return the new value
	}

	if blockNumber > -1 {
		msg = BlockRequestMessage{blockNumber, peerID}
	}
	return msg
}

func verifyBlocks(state State) State {
	done = false
	for !done {
		block = state.Store[height]

		if block != nil {
			// TODO: verify block.Block using block.Commit
			verified = true

			if verified {

> 				block.Execute()

Reviewer: step of `block.Execute()` takes most of the time of state transition. Since state transition is serial. We may find a way to let `block.Execute()` tasks run async.

Author: Agree. Verifying and executing blocks are good candidates to be isolated in a separate routine but as stated above the idea is to focus on performance improvements after (if needed).

}

type PeerStats struct {
	Height         int64
	PendingRequest int64 // it can be list in case there are multiple outstanding requests per peer
}

type BlockInfo struct {
	Block  Block
	Commit Commit
	PeerID ID // a peer from which we downloading the corresponding Block and Commit
}
```

The `Controller` is initialized by providing initial height (`startHeight`) from which it will start downloading blocks from peers.

``` go
func NewControllerState(startHeight int64, executor BlockExecutor) ControllerState {
	state = ControllerState {}
    state.Height = startHeight
	state.Mode = ModeFastSync
	state.MaxRequestPending = startHeight - 1
    state.PendingRequestsNum = 0
    state.Executor = executor
    initialize state.PeerMap, state.FailedRequests and state.Store to empty data structures
    return state
}

func handleEvent(state ControllerState, event Event) (ControllerState, Message, TimeoutTrigger, Error) {
	msg = nil
	timeout = nil
	error = nil

	switch state.Mode {
	case ModeConsensus:
		switch event := event.(type) {
		case EventBlockRequest:
			msg = createBlockResponseMessage(state, event)
			return state, msg, timeout, error
		default:

> 			error = "Only respond to BlockRequests while in ModeConsensus!"

Reviewer 3: ```suggestion
			error = "While in ModeConsensus only respond to Block Requests"
```

What about the status requests? Where do we reply to those?

Author: It is part of the Heartbeat mechanism that is part of the Executor logic. I assumed that Executor send periodically to each peer StatusRequest message and expects StatusReport to arrive within response timeout.

Reviewer 3: What's in a name... "peer"?  :)
A node that had previously answered to the status request broadcast? Or the peers the switch knows about (here we don't know the peer fast sync status)? 
If the former, will Executor keep track of peers in order to know who didn't answer?

Author: A peer is a node with which we have open connection (at the switch level). Executor sends status requests to all peers and need to manage timeouts. Therefore, we probably need to have peer lists at the Executor level also to able to determine who didn't responded on time.
